src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs

```diff
@@ -100,10 +132,17 @@ public static Diagnostic Filter(Diagnostic d, int warningLevelOption, ReportDiag
             int warningLevelOption,
             ReportDiagnostic generalDiagnosticOption,
             IDictionary<string, ReportDiagnostic> specificDiagnosticOptions,
+            ImmutableDictionary<SyntaxTree, ImmutableDictionary<string, ReportDiagnostic>> perTreeDiagnosticOptions,
```

Comment:
note: this seems interesting, because it means every time someone changes a compilation, they need to update this map for the compilation options.  That seems painful (And something someone might easily miss having to do).  If they don't update this map, then the current syntax trees in the compilation may not find anything in here if/when that syntax tree changes.

Reply:
Do you have an alternative?

Reply:
I believe we've often allowed "resolvers" to be passed in for this sort of thing.  So, we have things like that MetadataReferenceResolver, XmlReferenceResolver, etc.

So we should likely have a resolver for this purpose.  Maybe SyntaxTreeDiagnosticOptionsResolver.  Then, ti's clear that this isn't just data passed in.  But is instead an indirection that the host passes in that will be utilized over time to answer questions needed by the compilation.

Thoughts?

Reply:
The failure case doesn't really seem different. It would still need to be keyed on syntax tree, so the failure case is they ask the resolver for the options for the syntax tree, which has changed, and they forgot to update the resolver.

Reply:
You don't provide a new resolver.  The same resolver is used, but each time you are called back with teh appropriate SyntaxTree.  This is a very different model and makes it clear that hte compiler is asking the host to answer the question about this specific tree.  Importantly, there's no need to update multiple compiler objects as you're moving a Compilation forward with new trees.

Reply:
By "update the resolver" I mean register the new syntax tree with the resolver.

Reply:
IDE already registers SyntaxTrees in a central location to map them to our concepts.  Importantly, that's completely independent of making compilations.   So the resolver approach just works.  Importantly, by providing a resolver you are forced to ask yourself: what do i do when i get called back with any particular syntax tree.  htat's not the case when passing this dictionary to a compilation initially.  There's nothing that indicates that this can get out of sync with the compilation and that you must be continually changing this as you call .ReplaceSyntaxTree.

Reply:
I see. How about an IImmutableDictionary? I could make a new type, but it would have all of the methods of IImmutableDictionary so it doesn't seem super useful.

Reply:
Sorry ireadonlydictionary

Reply:
I really think anything that represents 'Data', but is keyed on 'SyntaxTree' just isn't the right abstraction.  It just means we have this incredibly strange point of inconsistancy in the Compilation API.

Even if we allowed this, the only way to really make things make sense would be to have to update compilations in two steps.  First, replace your trees, then replace your options.  Seems *super* strange and unlike anything in the compilation today.

Could you let me know what the concern would be with the Resolver pattern?  If we didn't do anything like that *today* i could understand a lot better :)   However, since we've commonly used that pattern to say "hey... host... can you go figure this out on demand", it seems pretty appropriate to me here :)

Reply:
I don't really understand what you mean by the "Resolver" pattern. The resolvers are abstract classes where you put in a key and return a value. In this case the key would be the syntax tree, the value would be the diagnostics, which just sounds like an IReadOnlyDictionary to me.

Reply:
It would be much more like a ```Func<SyntaxTree, ...>```  The difference is that a dictionary represents actual *realized* data that must be provided *and then updated on every syntax tree change*.  A func (or Resolver) represents that the data doesn't need to eb provided up front, can be computed on demand, and does not need to change as the compilation is mutated.

Frankly, a Func would be fine.  Except we already hve a pattern for those Funcs, it's: Resolver.  

--

And, to make the point again: 

> Even if we allowed this [the dictionary approach], the only way to really make things make sense would be to have to update compilations in two steps. First, replace your trees, then replace your options. Seems super strange and unlike anything in the compilation today.

Otherwise, you legitimately have compilations where these syntax trees will be out of sync.

Reply:
> I don't really understand what you mean by the "Resolver" pattern. The resolvers are abstract classes where you put in a key and return a value.

Yes... that's "the pattern".  You put in a key, and get a value back.  That's what you need for your scenario, just that the key is a SyntaxTree.  However, like with all the resolvers, you can just take in a singleton, and it can stay the same as the compilation moves along.  That entirely remove sthe problem of having to update the compilation options when literally any tree changes (which is also somewhat of an odd and unfortunate thing for a host to have to do).

Reply:
You're describing an implementation of IReadOnly  dictionary that is mutated behind the scenes. Func<SyntaxTree,...> is literally the type of the indexer method.

Reply:
If you accept an IReadOnlyDictionary it is highly misleading.  Practically no one sees the usage of IDictionary or IReadOnlyDictionary and things it represents data that will be computed in that manner.  The thought will be: i need to create a dictionary, fill it with the data, and pass it over to the options object for ownership.  

Furthermore, IReadOnlyDictionary is a large interface.  What does "Keys" or "Values" mean here?  

That does not seem like the contract we want to convey here.